src/plugins/intel_cpu/src/nodes/eltwise.cpp

```diff
@@ -2318,6 +2318,41 @@ void Eltwise::initSupportedPrimitiveDescriptors() {
     outputPrecision = filterPrecision(outputPrecision, forcedPrec);
     } else {
 #endif
+#if defined(OV_CPU_WITH_SHL)
+    auto filterPrecision = [&](const ov::element::Type& prc, const ov::element::Type& forcedPrec) {
+        if (isBitwise(algorithm)) {
+            if (std::find(supportedPrecisions.begin(), supportedPrecisions.end(), prc) == supportedPrecisions.end()) {
+                OPENVINO_THROW("Eltwise node with name `", getName(), "` doesn't support ", prc, " precision.");
+            }
+            return prc;
+        }
+        return forcedPrec;
+    };
+
+    const bool useShl = !useJit;
+    if (useShl) {
+    // Use original output precision as a reference point since some eltwise algorithms have non-float inputs (i.e. EltwiseSelect)
+    ov::element::Type forcedPrec = getOriginalOutputPrecisionAtPort(0) == ov::element::f16 ? ov::element::f16 : ov::element::f32;
+    // SHL implementation supports only identical precisions on inputs/outputs so they are aligned it to highest one
+    if (ShlEltwiseExecutor::isEltwiseAlgorithmSupported(getAlgorithm())) {
+        for (size_t i = 0; i < getParentEdges().size(); i++) {
+            if (!getParentEdgeAt(i)->getParent()->isConstant()) {
+                if (getOriginalInputPrecisionAtPort(i).size() > forcedPrec.size()) {
+                    forcedPrec = getOriginalInputPrecisionAtPort(i);
+                }
+            }
+        }
+        if (!forcedPrec.is_real()) {
+            forcedPrec = ov::element::f32;
+        }
+    }
+
+    for (size_t i = 0; i < inputPrecisions.size(); i++) {
+        inputPrecisions[i] = filterPrecision(inputPrecisions[i], forcedPrec);
+    }
+    outputPrecision = filterPrecision(outputPrecision, forcedPrec);
+    } else {
```

Comment:
Thanks so much for the really detailed explanation! I get to know more about the things that OV is doing!

So maybe I should remove `f16` related part and add an `else` statement after `if (ShlEltwiseExecutor::isEltwiseAlgorithmSupported(getAlgorithm()))` to force a precision conversion for reference implementation too?

Reply:
> So maybe I should remove f16 related part

Correct

> add an else statement after if (ShlEltwiseExecutor::isEltwiseAlgorithmSupported(getAlgorithm())) to force a 
precision conversion for reference implementation too?

I think it should be probably so:
```
if (useShl) {
    // If the op is supported by SHL executor, SHL requires FP32 precision on all inputs and outputs
   // If the op is not supported by SHL executor, it will be executed by reference implementation which requires fp32 precision on all inputs and outputs too (except bitwise).
   auto filterPrecision = [&](const ov::element::Type& prc) {
        if (ShlEltwiseExecutor::isEltwiseAlgorithmSupported(getAlgorithm())) {
            return ov::element::f32;
        } else if (isBitwise(algorithm)) {
                if (std::find(supportedPrecisions.begin(), supportedPrecisions.end(), prc) == supportedPrecisions.end()) {
                    OPENVINO_THROW("Eltwise node with name `", getName(), "` doesn't support ", prc, " precision.");
                }
                return prc;
        } else {  // reference
            return ov::element::f32;
        }
    };

    for (size_t i = 0; i < inputPrecisions.size(); i++) {
        inputPrecisions[i] = filterPrecision(inputPrecisions[i]);
    }
    outputPrecision = filterPrecision(outputPrecision);
}
```